// vior/src/ref.js
import Util from './util.js'
import Dep from './dep.js'

export default class Ref {
    static createRef(_this, _obj, _deps = null, depTag = null) {
        let obj = _obj && Util.isPlainObject(_obj) ? Util.deepCopy(_obj) : _obj,
            nobj = Array.isArray(obj) ? [] : {},
            deps = _deps || new Dep(_this)
        
        if (Util.isPlainObject(obj)) {
            for (let k in obj) {
                let v = obj[k]
                if (Util.isPlainObject(v))
                    nobj[k] = Ref.createRef(_this, v, deps, depTag || k)
                else
                    nobj[k] = v
            }
        } else {
            return obj
        }
        
        return new Proxy({
            __isViorRef: true,
            __isArray: Array.isArray(obj),
            __viorInstance: _this,
            __rawValue: obj,
            __realValue: nobj,
            __deps: deps,
            __depTag: depTag
        }, {
            get(target, key) {
                if (key == '__getRaw')
                    return (k = '__rawValue') => target[k]
                if (key == '__setRaw') {
                    return (k, v) => {
                        target.__rawValue[k] = v
                        target.__realValue[k] = Ref.createRef(_this, v, deps, depTag)
                    }
                }
                
                target.__deps.add(target.__depTag || key)
                return target.__realValue[key]
            },
            getOwnPropertyDescriptor(target, key) {
                return {
                    enumerable: true,
                    configurable: true
                }
            },
            ownKeys(target) {
                return Object.keys(target.__rawValue)
            },
            has(target, key) {
                return key in target.__rawValue
            },
            set(target, key, value) {
                let changed = value !== target.__rawValue[key]
                target.__rawValue[key] = value
                target.__realValue[key] = Ref.createRef(_this, value, deps, depTag)
                if (changed)
                    target.__deps.notify(target.__depTag || key)
                return true
            },
            deleteProperty(target, key) {
                delete target.__rawValue[key], target.__realValue[key]
                target.__deps.notify(target.__depTag || key)
                return true
            }
        })
    }
    static isRef(obj) {
        return obj && typeof obj.__getRaw == 'function' && obj.__getRaw('__isViorRef')
    }
    static isArrayRef(obj) {
        return this.isRef(obj) && obj.__getRaw('__isArray')
    }
}
